### PR TITLE
Fix Databricks cluster bootstrapping errors: Add UDRs for databricks

### LINKED ISSUE #141: CI build Databricks (and other) transient errors
We are seeing several transient errors which we will record here for tracking. We should consider finding root cause resolution and failing that use [Terragrunt error hooks](https://terragrunt.gruntwork.io/docs/features/hooks/) to retry them automatically.

### Cannot create cluster
```
Error: cannot create cluster: 0303-131037-qid874qx is not able to transition from TERMINATED to RUNNING: Self-bootstrap failure during launch. Please try again later and contact Databricks if the problem persists. databricks_error_message: Self-bootstrap failure during launch. Please try again later a..., Termination info: code: SELF_BOOTSTRAP_FAILURE, type: , parameters: map[databricks_error_message:Self-bootstrap failure during launch. Please try again later and contact Databricks if the problem persists. Node daemon fast failed and did not answer ping for instance 04387a5e9e4a4328b335937bb7db2e10]. Please see https://docs.databricks.com/dev-tools/api/latest/clusters.html#clusterclusterstate for more details
[2081](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2082)
│ 
[2082](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2083)
│   with databricks_cluster.fixed_single_node,
[2083](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2084)
│   on main.tf line 53, in resource "databricks_cluster" "fixed_single_node":
[2084](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2085)
│   53: resource "databricks_cluster" "fixed_single_node" {
```

Getting this often in new deployments to `westeurope`

***
+1 on the above error (uksouth)

### PR DESCRIPTION
**Pull request by jjgriff93**
# Resolves #141

## What is being addressed

This (hopefully) fixes the repeated issues we've seen when trying to deploy clusters. Following some investigation with the help of a support engineer, the reason why we might be seeing this only in the uksouth region currently is because each region has different IPs for various dependencies Databricks relies on and the routing might be clashing with other resources we have set up.

## How is this addressed

As per https://learn.microsoft.com/en-gb/azure/databricks/administration-guide/cloud-configurations/azure/udr?WT.mc_id=Portal-Microsoft_Azure_Support#--configure-user-defined-routes-with-azure-service-tags I've added a routing table with some service tags for these dependencies to ensure the routing happens correctly each time.

In my testing I've encountered no issues in uksouth however only time will tell.

### CODE REVIEW

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4534108852 (with refid `5ec2953b`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/213#issuecomment-1485414479) from @jjgriff93)

**Comment by jjgriff93** [comment]
Looks like the celebration was premature :( still getting an error on this PR run. Shall continue to investigate and move this to draft

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4544257482 (with refid `5ec2953b`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/213#issuecomment-1487001885) from @jjgriff93)

File: infrastructure/transform/network.tf [hunk 1]
```diff
@@ -69,12 +69,64 @@ resource "azurerm_subnet_network_security_group_association" "databricks_host" {
...
+
 resource "azurerm_private_dns_zone" "databricks" {
   name                = "privatelink.azuredatabricks.net"
   resource_group_name = var.core_rg_name
   tags                = var.tags
 }
 
+# Need for proper DNS resolution between Databricks & platform services
+# https://learn.microsoft.com/en-us/azure/virtual-network/what-is-ip-address-168-63-129-16
+resource "azurerm_private_dns_zone" "azure_platform" {
```
Comment (t-young31) [inline infrastructure/transform/network.tf:124]: think this should get [defined in core](https://github.com/UCLH-Foundry/FlowEHR/blob/aa73f432ec524fb4ca8c9e8dc4c4c39cbbe7aee1/infrastructure/core/locals.tf#L25) so it gets applied before any databricks resources

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4544775967 (with refid `5ec2953b`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/213#issuecomment-1487090480) from @jjgriff93)

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4547327795 (with refid `5ec2953b`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/213#issuecomment-1487547172) from @jjgriff93)

**Pull request closed without merging**